### PR TITLE
Webhook v2 job type

### PR DESCRIPTION
**Pull request by spooktheducks**

### CODE REVIEW

**Review by archseer — APPROVED** [review]
Needs a fix in the CHANGELOG, but can be done in a follow-up to unblock the merge

File: core/services/job/models.go [hunk 1]
```diff
@@ -142,10 +144,45 @@ func (OffchainReportingOracleSpec) TableName() string {
...
+}
+
+func (w WebhookSpec) GetID() string {
+	return fmt.Sprintf("%v", w.ID)
+}
+
+func (w *WebhookSpec) SetID(value string) error {
+	ID, err := strconv.ParseInt(value, 10, 32)
+	if err != nil {
+		return err
```
Comment (PiotrTrzpil) [inline core/services/job/models.go:161]: minor, but could be wrapped with some context

File: core/services/pipeline/runner.go [hunk 2]
```diff
@@ -137,12 +136,35 @@ func (m *memoryTaskRun) results() (a []Result) {
 	return
 }
 
+func (m *memoryTaskRun) numPredecessorsRemaining() int {
+	m.predMu.RLock()
+	defer m.predMu.RUnlock()
+	return m.predecessorsRemaining
+}
+
+func (m *memoryTaskRun) tryToClaim() (ok bool) {
```
Comment (PiotrTrzpil) [inline core/services/pipeline/runner.go:145]: hm it's a bit non-obvious what claiming has to do with `m.finished` - maybe `tryFinish` or `trySwitchToFinished` ?

File: core/services/pipeline/runner.go [hunk 3]
```diff
@@ -336,27 +309,70 @@ func (r *runner) executeRun(ctx context.Context, txdb *gorm.DB, spec Spec, meta
...
 	l.Debugw("Finished all tasks for pipeline run", "specID", spec.ID, "runTime", runTime)
 	promPipelineRunTotalTimeToCompletion.WithLabelValues(fmt.Sprintf("%d", spec.JobID), spec.JobName).Set(float64(runTime))
-	if retry || trrs.FinalResult().HasErrors() {
+	if retry || taskRunResults.FinalResult().HasErrors() {
 		promPipelineRunErrors.WithLabelValues(fmt.Sprintf("%d", spec.JobID), spec.JobName).Inc()
 	}
 	if !retry {
-		run.Errors = trrs.FinalResult().ErrorsDB()
-		run.Outputs = trrs.FinalResult().OutputsDB()
+		run.Errors = taskRunResults.FinalResult().ErrorsDB()
```
Comment (PiotrTrzpil) [inline core/services/pipeline/runner.go:316]: yup, trrs is not ideal ;)

File: core/services/pipeline/runner.go [hunk 4]
```diff
@@ -336,27 +309,70 @@ func (r *runner) executeRun(ctx context.Context, txdb *gorm.DB, spec Spec, meta
...
 	}
-	return run, trrs, retry, err
+	return run, taskRunResults, retry, err
 }
 
-func (r *runner) executeTaskRun(ctx context.Context, spec Spec, task Task, meta JSONSerializable, inputs []Result, l logger.Logger) Result {
+func (r *runner) memoryTaskRunDAGFromTaskDAG(taskDAG TaskDAG, pipelineInputs []Result, txdb *gorm.DB) (headTaskRuns []*memoryTaskRun, _ error) {
+	tasks, err := taskDAG.TasksInDependencyOrder()
+	if err != nil {
+		return nil, err
```
Comment (PiotrTrzpil) [inline core/services/pipeline/runner.go:326]: wrap in some context? e.g. 'Failed to arrange tasks in dependency order'

File: core/services/webhook/delegate.go [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+	defer r.muSpecsByUUID.RUnlock()
+	spec, exists := r.specsByUUID[jobID]
+	return spec, exists
+}
+
+var ErrJobNotExists = errors.New("job does not exist")
+
+func (r *webhookJobRunner) RunJob(ctx context.Context, jobUUID models.JobID, pipelineInputs []pipeline.Result, meta pipeline.JSONSerializable) (int64, error) {
+	var uuidHash models.JobID
+	copy(uuidHash[:], jobUUID[:])
```
Comment (PiotrTrzpil) [inline core/services/webhook/delegate.go:110]: maybe worth creating method `jobUUID.Copy()` ? but also, why is the copy needed?

File: core/services/webhook/delegate.go [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+		),
+	)
+
+	ctx, cancel := utils.CombinedContext(ctx, spec.chRemove)
+	defer cancel()
+
+	runID, _, err := r.runner.ExecuteAndInsertFinishedRun(ctx, *spec.PipelineSpec, pipelineInputs, meta, *logger, true)
+	if err != nil {
+		logger.Errorw("Error running pipeline for webhook job", "error", err)
+		return 0, err
```
Comment (PiotrTrzpil) [inline core/services/webhook/delegate.go:130]: could that lead to logging twice? might be good to wrap the error instead?

File: core/services/webhook/delegate.go (hunk 5 above)
Comment (archseer) [reply core/services/webhook/delegate.go:110]: I was wondering about this too, I thought a cast might have worked

**Pull request merged**